File: scripts/web_search.py
# -*- coding: utf-8 -*-
import hashlib
import hmac
import json
import sys
import time
from datetime import datetime
from typing import List, Optional
import logging

import requests

logging.basicConfig(level=logging.DEBUG)


def _parse_response(response):
    resp_list = []
    if "Response" in response:
        data = response["Response"]
        if "Pages" in data:
            webPages = data["Pages"]
            # TODO 待处理
            num = 0
            for item_str in webPages:
                num += 1
                item = json.loads(item_str)
                resp_list.append(
                    {
                    "id": item.get("id", num),
                    "title": item.get("title", ""),
                    "url": item.get("url", ""),
                    "summary": item.get("passage", ""),
                    "detail": item.get("passage", ""),
                    "siteName": item.get("siteName", ""),
                    "siteIcon": item.get("favicon", ""),
                    "datePublished": item.get("date", ""),
                    "score": item.get("score", 0)
                    }
                )
            resp_list.sort(key=lambda item: item.get("score"), reverse=True)
    return resp_list


def sign(key, msg):
    return hmac.new(key, msg.encode("utf-8"), hashlib.sha256).digest()

# 实例化一个认证对象，入参需要传入腾讯云账户 SecretId 和 SecretKey，此处还需注意密钥对的保密
# 代码泄露可能会导致 SecretId 和 SecretKey 泄露，并威胁账号下所有资源的安全性。以下代码示例仅供参考，建议采用更安全的方式来使用密钥，请参见：https://cloud.tencent.com/document/product/1278/85305
# 密钥可前往官网控制台 https://console.cloud.tencent.com/cam/capi 进行获取


import os


def search_tencent(
    query: str, 
    secret_id: str = os.environ.get("TENCENT_SECRET_ID"), 
    secret_key: str= os.environ.get("TENCENT_SECRET_KEY"), 
    count: int = 10
):
    """Search using tencent's Search API and return the results as a list.

    Args:
        secret_id (str): A tencent Search secret_id
        secret_key (str): A tencent Search secret_key
        query (str): The query to search for
    """
    token = ""

    service = "tms"
    host = "tms.tencentcloudapi.com"
    region = ""
    version = "2020-12-29"
    action ="SearchPro"  #  "TextModeration"
    payload_dict: dict = {
        "Query": query,
        # "Version": version,
        # "Action": action
        }
    payload = json.dumps(payload_dict)
    
    # params = json.loads(payload)
    # endpoint = "https://tms.tencentcloudapi.com"
    algorithm = "TC3-HMAC-SHA256"
    timestamp = int(time.time())
    date = datetime.utcfromtimestamp(timestamp).strftime("%Y-%m-%d")

    # ************* 步骤 1：拼接规范请求串 *************
    http_request_method = "POST"
    canonical_uri = "/"
    canonical_querystring = ""
    ct = "application/json; charset=utf-8"
    canonical_headers = "content-type:%s\nhost:%s\nx-tc-action:%s\n" % (ct, host, action.lower())
    signed_headers = "content-type;host;x-tc-action"
    hashed_request_payload = hashlib.sha256(payload.encode("utf-8")).hexdigest()
    canonical_request = (http_request_method + "\n" +
                        canonical_uri + "\n" +
                        canonical_querystring + "\n" +
                        canonical_headers + "\n" +
                        signed_headers + "\n" +
                        hashed_request_payload)

    # ************* 步骤 2：拼接待签名字符串 *************
    credential_scope = date + "/" + service + "/" + "tc3_request"
    hashed_canonical_request = hashlib.sha256(canonical_request.encode("utf-8")).hexdigest()
    string_to_sign = (algorithm + "\n" +
                    str(timestamp) + "\n" +
                    credential_scope + "\n" +
                    hashed_canonical_request)

    # ************* 步骤 3：计算签名 *************
    secret_date = sign(("TC3" + secret_key).encode("utf-8"), date)
    secret_service = sign(secret_date, service)
    secret_signing = sign(secret_service, "tc3_request")
    signature = hmac.new(secret_signing, string_to_sign.encode("utf-8"), hashlib.sha256).hexdigest()

    # ************* 步骤 4：拼接 Authorization *************
    authorization = (algorithm + " " +
                    "Credential=" + secret_id + "/" + credential_scope + ", " +
                    "SignedHeaders=" + signed_headers + ", " +
                    "Signature=" + signature)

    # ************* 步骤 5：构造并发起请求 *************
    headers = {
        "Authorization": authorization,
        "Content-Type": "application/json; charset=utf-8",
        "Host": host,
        "X-TC-Action": action,
        "X-TC-Timestamp": str(timestamp),
        "X-TC-Version": version
    }
    if region:
        headers["X-TC-Region"] = region
    if token:
        headers["X-TC-Token"] = token

    try:

        response = requests.post("https://" + host, headers=headers, data=payload.encode("utf-8"))
        response.raise_for_status()
        return _parse_response(response.json())
    except Exception as err:
        print(err)

from typing import Dict, List
import requests
import json


def network_retrieval_bocha(query):
    """
    联网搜索，返回AI搜索结果列表。
    :param query: str,
    :return: List[dict],
    """
    url = "https://api.bochaai.com/v1/ai-search"
    payload = json.dumps({
        "query": f"{query}",
        "freshness": "oneYear",
        "count": 50,
        "answer": False,
        "stream": False,
    })
    headers = {
        'Authorization': 'sk-5caab5d8495f4ac08e5eda4e34802816',
        'Content-Type': 'application/json',
        'Connection': 'keep-alive',
    }
    response = requests.request("POST", url, headers=headers, data=payload)
    txt = response.json()
    res = json.loads(txt['messages'][0]['content'])['value']
    return res


##
def rerank_bocha(query, documents: list):
    """
        语义重排序，返回Top-N排序索引与分数。
        :param query: str,
        :param documents: List[str],
        :return: List[tuple],
        """
    url = "https://api.bochaai.com/v1/rerank"
    payload = json.dumps({
        "model": "gte-rerank",
        "query": query,
        "documents": documents,
        "return_documents": True,
    })
    headers = {
        'Authorization': 'sk-5caab5d8495f4ac08e5eda4e34802816',
        'Content-Type': 'application/json',
    }
    response = requests.request("POST", url, headers=headers, data=payload)
    if response and response.status_code == 200 and 'data' in response.json():
        return [(info['index'], info['relevance_score']) for info in response.json()['data']['results']]
    else:
        return []


def rerank(query, documents: List[str], web_pages: List[Dict]):
    """
    语义重排序，返回Top-N排序索引与分数。
    :param query: str,
    :param documents: List[str], web_pages 中的 summary
    :param exist_corpus: List[str], 从知识库检索到的语料
    :param web_pages: List[Dict], 从网络（如腾讯）检索到的网页
    :return: List[tuple],
    """
    ranked_pages = rerank_bocha(query, documents)
    show_pages = []
    for index, score in ranked_pages:
        if score < 0.4:
            break
        show_pages.append(web_pages[index])
    return show_pages


def search_web(query: str) -> str:
    tencent_results = search_tencent(query) # search_tencent(keywords + " " + query)

    documents = [item.get("summary", "") for item in tencent_results]
    reranked_results = rerank(query, documents, tencent_results)
    return "\n".join([r.get("summary") for r in reranked_results]), reranked_results


if __name__ == "__main__":
    resp_summary, resp_list = search_web(query="使用sharding-jdbc进行分库分表，如何结合用户ID(user_id)和时间范围进行分片")
    print(resp_summary)

File: server/utils/webSearch.js

import crypto from 'crypto';
import fetch from 'node-fetch';

function sign(key, msg) {
    return crypto.createHmac('sha256', key).update(msg).digest();
}

function parseResponse(response) {
    let respList = [];
    if (response && response.Response) {
        const data = response.Response;
        if (data.Pages) {
            const webPages = data.Pages;
            let num = 0;
            for (const itemStr of webPages) {
                num += 1;
                // itemStr is a JSON string in the python script, but let's check if it's already an object here. 
                // The python script says `item = json.loads(item_str)`. 
                // If the API returns a list of strings, we parse them. If it returns objects, we use them.
                // Looking at Tencent API docs (or assuming based on python), it seems to be a list of strings if `json.loads` is used.
                let item;
                try {
                    item = typeof itemStr === 'string' ? JSON.parse(itemStr) : itemStr;
                } catch (e) {
                    console.error('Error parsing itemStr:', e);
                    continue;
                }

                respList.push({
                    id: item.id || num,
                    title: item.title || "",
                    url: item.url || "",
                    summary: item.passage || "",
                    detail: item.passage || "",
                    siteName: item.siteName || "",
                    siteIcon: item.favicon || "",
                    datePublished: item.date || "",
                    score: item.score || 0
                });
            }
            respList.sort((a, b) => b.score - a.score);
        }
    }
    return respList;
}

async function searchTencent(query, count = 10) {
    const secretId = process.env.TENCENT_SECRET_ID;
    const secretKey = process.env.TENCENT_SECRET_KEY;

    if (!secretId || !secretKey) {
        throw new Error("Tencent API credentials not found.");
    }

    const service = "tms";
    const host = "tms.tencentcloudapi.com";
    const region = "";
    const version = "2020-12-29";
    const action = "SearchPro";
    const payloadDict = {
        "Query": query
    };
    const payload = JSON.stringify(payloadDict);

    const algorithm = "TC3-HMAC-SHA256";
    const timestamp = Math.floor(Date.now() / 1000);
    const date = new Date(timestamp * 1000).toISOString().split('T')[0];

    // ************* Step 1: Canonical Request String *************
    const httpRequestMethod = "POST";
    const canonicalUri = "/";
    const canonicalQuerystring = "";
    const ct = "application/json; charset=utf-8";
    const canonicalHeaders = `content-type:${ct}\nhost:${host}\nx-tc-action:${action.toLowerCase()}\n`;
    const signedHeaders = "content-type;host;x-tc-action";
    const hashedRequestPayload = crypto.createHash('sha256').update(payload).digest('hex');
    const canonicalRequest = [
        httpRequestMethod,
        canonicalUri,
        canonicalQuerystring,
        canonicalHeaders,
        signedHeaders,
        hashedRequestPayload
    ].join('\n');

    // ************* Step 2: String to Sign *************
    const credentialScope = `${date}/${service}/tc3_request`;
    const hashedCanonicalRequest = crypto.createHash('sha256').update(canonicalRequest).digest('hex');
    const stringToSign = [
        algorithm,
        timestamp.toString(),
        credentialScope,
        hashedCanonicalRequest
    ].join('\n');

    // ************* Step 3: Signature *************
    const secretDate = sign(Buffer.from("TC3" + secretKey, 'utf-8'), date);
    const secretService = sign(secretDate, service);
    const secretSigning = sign(secretService, "tc3_request");
    const signature = crypto.createHmac('sha256', secretSigning).update(stringToSign).digest('hex');

    // ************* Step 4: Authorization Header *************
    const authorization = `${algorithm} Credential=${secretId}/${credentialScope}, SignedHeaders=${signedHeaders}, Signature=${signature}`;

    // ************* Step 5: Request *************
    const headers = {
        "Authorization": authorization,
        "Content-Type": "application/json; charset=utf-8",
        "Host": host,
        "X-TC-Action": action,
        "X-TC-Timestamp": timestamp.toString(),
        "X-TC-Version": version
    };
    if (region) headers["X-TC-Region"] = region;

    try {
        const response = await fetch(`https://${host}`, {
            method: 'POST',
            headers: headers,
            body: payload
        });

        if (!response.ok) {
            const errorText = await response.text();
            throw new Error(`Tencent API request failed with status ${response.status}: ${errorText}`);
        }

        const data = await response.json();
        return parseResponse(data);
    } catch (err) {
        console.error("Tencent Search Error:", err);
        return [];
    }
}

async function rerankBocha(query, documents) {
    const url = "https://api.bochaai.com/v1/rerank";
    const apiKey = process.env.BOCHA_API_KEY; // Fallback to key from python script

    const payload = JSON.stringify({
        "model": "gte-rerank",
        "query": query,
        "documents": documents,
        "return_documents": true,
    });

    const headers = {
        'Authorization': apiKey,
        'Content-Type': 'application/json',
    };

    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: headers,
            body: payload
        });

        if (response.ok) {
            const data = await response.json();
            if (data && data.data && data.data.results) {
                return data.data.results.map(info => ({
                    index: info.index,
                    score: info.relevance_score
                }));
            }
        }
    } catch (error) {
        console.error("Bocha Rerank Error:", error);
    }
    return [];
}

async function rerank(query, documents, webPages) {
    const rankedPagesMetadata = await rerankBocha(query, documents);
    const showPages = [];

    // Convert rerank results to look like the python loop:
    // for index, score in ranked_pages:
    //     if score < 0.4: break
    //     show_pages.append(web_pages[index])

    // Note: Bocha API might assume documents list order is preserved in index, 
    // so `documents` passed to rerank must match `webPages` order.

    // Sort logic from python script implies that `rerankBocha` returns sorted results?
    // The python script just iterates over the returned list directly. 
    // The Bocha API response structure: `results` is a list of objects with `index` and `relevance_score`.
    // Usually rerank APIs return them sorted by score descending.

    for (const info of rankedPagesMetadata) {
        if (info.score < 0.4) {
            // Logic in python script: `break`.
            // Depending on if the API returns sorted results, this truncates the tail.
            // If not sorted, this might skip good results that just happened to be later in list?
            // Assuming sorted.
            break; // Python: break
        }
        if (webPages[info.index]) {
            showPages.push(webPages[info.index]);
        }
    }
    return showPages;
}

export async function searchWeb(query) {
    // Python: tencent_results = search_tencent(query)
    const tencentResults = await searchTencent(query);

    // Python: documents = [item.get("summary", "") for item in tencent_results]
    const documents = tencentResults.map(item => item.summary || "");

    // Python: reranked_results = rerank(query, documents, tencent_results)
    const rerankedResults = await rerank(query, documents, tencentResults);

    // Python: return "\n".join([r.get("summary") for r in reranked_results]), reranked_results
    const summary = rerankedResults.map(r => r.summary).join("\n");

    return {
        summary,
        results: rerankedResults
    };
}
